### DISCUSSION TITLE
走docker容器时，socket5代理连接建立失败

### DISCUSSION THREAD

**Discussion by Shxuai** [docker-easyconnect/docker-easyconnect#62] [Q&A] [answered]
以下为我的启动脚本

```
[root@localhost ~]# docker run --device /dev/net/tun --cap-add NET_ADMIN --name="vpn" -ti -v $HOME/.easyconn:/root/.easyconn -p 127.0.0.1:1080:1080 -e EC_VER=7.6.3 hagb/docker-easyconnect:cli
Authenticate failed, reason: Invalid username or password!
user "**********" auto login failed
vpn adress: https://ssl.*******.com:4451
username: **********
password: **********
user "**********" login successfully!
```

telnet本机1080端口可通
但是curl的时候就会提示无法建立socket5连接
以下为crul的报错信息

```
[root@localhost ~]# curl --proxy socks5h://localhost:1080 http://10.80.21.229:8080/#/login -v
* About to connect() to proxy localhost port 1080 (#0)
*   Trying ::1...
* 拒绝连接
*   Trying 127.0.0.1...
* Can't complete SOCKS5 connection to 172.17.0.2:47688. (4)
* Closing connection 0
curl: (7) Can't complete SOCKS5 connection to 172.17.0.2:47688. (4)
```

**Comment by Hagb** [comment, marked as answer]
我在我这里没有复现出来这个问题（也没有想到可能的原因）。现在问题依然存在不？尝试下登录后在容器里面用`busybox wget http://10.80.21.229:8080/#/login` 能不能成功？

**Reply by Shxuai** [reply]
烦请问，你用的系统是什么版本呢？我用的centos7失败，8可以成功。不知道是不是系统兼容性问题。

**Reply by Hagb** [reply]
7 尝试加上 `-e IPTABLES_LEGACY=1` 参数？

**Reply by Shxuai** [reply]
以下为我的docker启动脚本

```
[root@vultr ~]# docker run --device /dev/net/tun --cap-add NET_ADMIN --name="vpn" -ti -v $HOME/.easyconn:/root/.easyconn -p 127.0.0.1:1080:1080 -e EC_VER=7.6.3 -e CLI_OPTS="-d https://ssl.******.com:4451 -u ****** -p ******" -e IPTABLES_LEGACY=1 hagb/docker-easyconnect:cli
Unable to find image 'hagb/docker-easyconnect:cli' locally
Trying to pull repository docker.io/hagb/docker-easyconnect ... 
cli: Pulling from docker.io/hagb/docker-easyconnect
69692152171a: Pull complete 
e503432c0730: Pull complete 
5f4c2667a60f: Pull complete 
2aa994f7c601: Pull complete 
a584890fe699: Pull complete 
921eefd37e14: Pull complete 
Digest: sha256:fb3a04bed5314deb514fb3e8596e5e5077ac303d35c51fc269e7288e3e116876
Status: Downloaded newer image for docker.io/hagb/docker-easyconnect:cli
update-alternatives: using /usr/sbin/iptables-legacy to provide /usr/sbin/iptables (iptables) in manual mode
update-alternatives: using /usr/sbin/ip6tables-legacy to provide /usr/sbin/ip6tables (ip6tables) in manual mode
iptables v1.8.2 (legacy): unknown option "--restore-mark"
Try `iptables -h' or 'iptables --help' for more information.
iptables v1.8.2 (legacy): Couldn't load match `connmark':No such file or directory

Try `iptables -h' or 'iptables --help' for more information.
iptables v1.8.2 (legacy): Couldn't load match `connmark':No such file or directory

Try `iptables -h' or 'iptables --help' for more information.
iptables v1.8.2 (legacy): unknown option "--set-mark"
Try `iptables -h' or 'iptables --help' for more information.
auto login is disabled
user "******" login successfully!
```

在curl请求时有新错误产生

```
[root@vultr ~]# curl --proxy socks5h://localhost:1080 http://10.80.21.229:8080/#/login -v
* About to connect() to proxy localhost port 1080 (#0)
*   Trying ::1...
* Connection refused
*   Trying 127.0.0.1...
* Unable to receive initial SOCKS5 response.
* Closing connection 0
curl: (7) Unable to receive initial SOCKS5 response.

```

**Reply by Hagb** [reply]
socks 服务会在 vpn 连接成功后的 0~5 秒启动，再试试 curl 看看行不行？

**Reply by Shxuai** [reply]
> socks 服务会在 vpn 连接成功后的 0~5 秒启动，再试试 curl 看看行不行？

可以了解决了，按照你说的7下面加一个`-e IPTABLES_LEGACY=1`参数。麻烦同步到文档，让更多人看到。

**Reply by Hagb** [reply]
我想想办法能不能自动检测……

**Reply by Hagb** [reply]
> 我想想办法能不能自动检测……

@Shxuai 方便在不加 `-e IPTABLES_LEGACY=1` 参数的情况下试一下 `hagb/docker-easyconnect:cli-dev` 镜像吗？

**Reply by Shxuai** [reply]
现在无论cli和cli-dev都会有以下报错
![image](https://user-images.githubusercontent.com/26001241/126172623-5724fed2-3302-49fc-843e-d587217efcbc.png)

**Reply by Hagb** [reply]
> 现在无论cli和cli-dev都会有以下报错
> ![image](https://user-images.githubusercontent.com/26001241/126172623-5724fed2-3302-49fc-843e-d587217efcbc.png)

是不是原来运行着的容器没退出？

**Reply by Shxuai** [reply]
我是全新虚拟的一台服务器，而且也检查过了容器全删除了。我再试一次。

**Reply by Hagb** [reply]
刚刚查了查 https://blog.csdn.net/fly910905/article/details/98809355
试试重启 docker 服务？
``` bash
systemctl restart docker
````

**Reply by Shxuai** [reply]
依然没有成功
`[root@vultr ~]# touch ~/.easyconn
[root@vultr ~]# docker run --device /dev/net/tun --cap-add NET_ADMIN -ti -v $HOME/.easyconn:/root/.easyconn -p 127.0.0.1:1080:1080 -e EC_VER=7.6.3 hagb/docker-easyconnect:cli-dev
Unable to find image 'hagb/docker-easyconnect:cli-dev' locally
Trying to pull repository docker.io/hagb/docker-easyconnect ... 
cli-dev: Pulling from docker.io/hagb/docker-easyconnect
b4d181a07f80: Pull complete 
45f868565b65: Pull complete 
54960605877d: Pull complete 
3f24b5fff659: Pull complete 
cc48e48e0fd5: Pull complete 
9a28eb6d98b1: Pull complete 
Digest: sha256:9a84fdec487a47dbbe0bbb5ced7a34ee43711d6bc571a45cd36f4c287f73435d
Status: Downloaded newer image for docker.io/hagb/docker-easyconnect:cli-dev
open: No such device
Failed to operate tun device! Please check whether /dev/net/tun is available.
[root@vultr ~]# service firewalld stop
Redirecting to /bin/systemctl stop firewalld.service
[root@vultr ~]# docker run --device /dev/net/tun --cap-add NET_ADMIN -ti -v $HOME/.easyconn:/root/.easyconn -p 127.0.0.1:1080:1080 -e EC_VER=7.6.3 hagb/docker-easyconnect:cli-dev
/usr/bin/docker-current: Error response from daemon: driver failed programming external connectivity on endpoint relaxed_jang (778c0a9471db20adc3bfe111e0f3c9b73cfe3251f2427be25076b6ad213e23ee):  (iptables failed: iptables --wait -t nat -A DOCKER -p tcp -d 127.0.0.1 --dport 1080 -j DNAT --to-destination 172.17.0.2:1080 ! -i docker0: iptables: No chain/target/match by that name.
 (exit status 1)).
[root@vultr ~]# systemctl restart docker
[root@vultr ~]# docker run --device /dev/net/tun --cap-add NET_ADMIN -ti -v $HOME/.easyconn:/root/.easyconn -p 127.0.0.1:1080:1080 -e EC_VER=7.6.3 hagb/docker-easyconnect:cli-dev
open: No such device
Failed to operate tun device! Please check whether /dev/net/tun is available.
[root@vultr ~]# 
`

**Reply by Shxuai** [reply]
我可以提供ssh方式给你，方便你测试。服务器是临时租赁的，麻烦测试结束通知一下我释放服务器。
149.28.79.119
root
b8_Ewk]u69ChELMh

**Reply by Hagb** [reply]
感谢！帮大忙了。各种不同环境的情况还是不太一样…… 等 docker hub 构建完镜像我再试试。

**Reply by Shxuai** [reply]
> 感谢！帮大忙了。各种不同环境的情况还是不太一样…… 等 docker hub 构建完镜像我再试试。
那服务器用完了吗？如果用完了我先释放了，这个收费的。

**Reply by Hagb** [reply]
用完了。
现在应该可以了。方便再测试一下不？
可能会出现 `Can't find available method to automatically set route for opening ports (refer to https://github.com/Hagb/docker-easyconnect/tree/master/doc/route.md)` 的报错。如果不将端口开放给外部，这个报错忽略就好。
```bash
modprobe tun
docker run --device /dev/net/tun --cap-add NET_ADMIN -ti -v $HOME/.easyconn:/root/.easyconn -p 127.0.0.1:1080:1080 -e EC_VER=7.6.3 hagb/docker-easyconnect:cli-dev
```

**Reply by Shxuai** [reply]
好的那我重装一下服务器，在测试一次

**Reply by Shxuai** [reply]
可以使用了，十分感谢！
